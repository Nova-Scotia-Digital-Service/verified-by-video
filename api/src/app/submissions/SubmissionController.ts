import * as TD from '../../types'

import {
  Get,
  Controller,
  Param,
  Post,
  HttpException,
  HttpStatus,
  UseInterceptors,
  UploadedFile,
  Body,
  UseGuards,
  Delete,
} from '@nestjs/common'
import { FileInterceptor } from '@nestjs/platform-express'
import { ApiConsumes, ApiProperty, ApiBearerAuth as SwaggerRequireAuth } from '@nestjs/swagger'
import { v4 as uuid } from 'uuid'

import config from '../../config'
import { minioClient } from '../../minio'

import { AuthGuard } from '../auth/AuthGuard'

import { createPhotoID, createSubmission, getSubmission, getSubmissionList } from './SubmissionData'
import { applyTagToSubmission, removeTagFromSubmission } from '../tags/TagData'

class SubmissionBodySchema {
  @ApiProperty({ type: 'string' })
  sessionId!: string

  @ApiProperty({ type: 'string', format: 'binary' })
  video_file!: Express.Multer.File
}

class PhotoDto {
  @ApiProperty({ type: 'string' })
  sessionId!: string

  @ApiProperty({ type: 'string' })
  description!: string

  @ApiProperty({ type: 'string', format: 'binary' })
  photo_file!: Express.Multer.File
}

@Controller('/submissions')
export class SubmissionController {
  @Post('/')
  @ApiConsumes('multipart/form-data')
  @UseInterceptors(FileInterceptor('video_file'))
  public async createSubmission(
    @Body() { sessionId: session_id }: SubmissionBodySchema,
    @UploadedFile() video_file: Express.Multer.File,
  ): Promise<TD.Submission> {
    const fileNameComponents = video_file.originalname.split('.')
    const fileExtension = fileNameComponents[fileNameComponents.length - 1]
    const randomizedFileName = `${uuid()}.${fileExtension}`
    const filePath = `media/${randomizedFileName}`

    await minioClient.putObject(config.S3_BUCKET_NAME, filePath, video_file.buffer, video_file.size)

    const submission = await createSubmission(session_id, filePath)
    return submission
  }

  @Post('/photo')
  @ApiConsumes('multipart/form-data')
  @UseInterceptors(FileInterceptor('photo_file'))
  public async createPhotoID(
    @Body() { sessionId: session_id, description }: PhotoDto,
    @UploadedFile() photo_file: Express.Multer.File,
  ): Promise<TD.IdentificationCard> {
    const fileNameComponents = photo_file.originalname.split('.')
    const fileExtension = fileNameComponents[fileNameComponents.length - 1]
    const randomizedFileName = `${uuid()}.${fileExtension}`
    const filePath = `media/${randomizedFileName}`

    await minioClient.putObject(config.S3_BUCKET_NAME, filePath, photo_file.buffer, photo_file.size)

    return await createPhotoID(session_id, description, filePath)
  }

  @Get('/:submissionId')
  public async getSubmission(@Param('submissionId') submission_id: string): Promise<TD.Submission> {
    const submission = await getSubmission(submission_id)
    if (submission === undefined) {
      throw new HttpException('Not found', HttpStatus.NOT_FOUND)
    }
    return submission
  }
}

@Controller('/staff/submissions')
export class StaffSubmissionController {
  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Get('/')
  public async getSubmissionList(): Promise<TD.APISubmissionSummary[]> {
    const submissionList = await getSubmissionList()

    return submissionList.map((submission) => ({
      ...submission,
      reviews: submission.reviews.map((review) => ({
        ...review,
        created_at: new Date(review.created_at),
      })),
    }))
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Post('/:submission_id/tag')
  public async applyTagToSubmission(
    @Param('submission_id') submission_id: string,
    @Body() { tag_id }: { tag_id: string },
  ): Promise<void> {
    await applyTagToSubmission(submission_id, tag_id)
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Delete('/:submission_id/tag/:tag_id')
  public async removeTagFromSubmission(
    @Param('submission_id') submission_id: string,
    @Param('tag_id') tag_id: string,
  ): Promise<void> {
    await removeTagFromSubmission(submission_id, tag_id)
  }
}
