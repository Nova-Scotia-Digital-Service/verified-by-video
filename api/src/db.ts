import { Pool } from 'pg'

import config from './config'

export const pool = new Pool({
  user: config.PG_USER,
  password: config.PG_PASSWORD,
  host: config.PG_HOST,
  database: config.PG_DATABASE,
})
