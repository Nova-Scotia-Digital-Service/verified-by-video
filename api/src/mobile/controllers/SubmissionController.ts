import { Get, JsonController, Param, Post } from 'routing-controllers'
import { Service } from 'typedi'

import { confirmSubmission, previousSubmission } from '../content/SubmissionData'

@JsonController('/submissions')
@Service()
export class SubmissionController {
  @Get('/') // TEMPORARY - for development
  @Post('/')
  public async postSubmissions() {
    return confirmSubmission
  }

  @Get('/:submissionId')
  public async getSubmission(@Param('submissionId') submission_id: string) {
    return {
      ...previousSubmission,
      id: submission_id,
    }
  }
}
