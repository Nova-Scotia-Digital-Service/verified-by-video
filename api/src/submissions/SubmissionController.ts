import * as TD from '../types'

import {
  Get,
  Controller,
  Param,
  Post,
  HttpException,
  HttpStatus,
  UseInterceptors,
  UploadedFile,
  Body,
} from '@nestjs/common'

import { FileInterceptor } from '@nestjs/platform-express'
import { ApiConsumes, ApiProperty } from '@nestjs/swagger'
import { v4 as uuid } from 'uuid'

import { createSubmission, getSubmission } from './SubmissionData'

import config from '../config'
import { minioClient } from '../minio'

class SubmissionDto {
  @ApiProperty({ type: 'string' })
  sessionId!: string

  @ApiProperty({ type: 'string', format: 'binary' })
  file!: Express.Multer.File
}

@Controller('/submissions')
export class SubmissionController {
  @Post('/')
  @ApiConsumes('multipart/form-data')
  @UseInterceptors(FileInterceptor('file'))
  public async createSubmission(
    @Body() { sessionId: session_id }: SubmissionDto,
    @UploadedFile() file: Express.Multer.File,
  ): Promise<TD.Submission> {
    const fileNameComponents = file.originalname.split('.')
    const fileExtension = fileNameComponents[fileNameComponents.length - 1]
    const randomizedFileName = `${uuid()}.${fileExtension}`
    const filePath = `media/${randomizedFileName}`

    await minioClient.putObject(config.S3_BUCKET_NAME, filePath, file.buffer, file.size)

    return await createSubmission(session_id, filePath)
  }

  @Get('/:submissionId')
  public async getSubmission(@Param('submissionId') submission_id: string): Promise<TD.Submission> {
    const submission = await getSubmission(submission_id)
    if (submission === undefined) {
      throw new HttpException('Not found', HttpStatus.NOT_FOUND)
    }
    return submission
  }
}
