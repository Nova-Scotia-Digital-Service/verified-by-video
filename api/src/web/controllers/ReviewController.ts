import { Get, JsonController } from 'routing-controllers'
import { Service } from 'typedi'

import { videoReview } from '../content/VideoData'

@JsonController('/reviews')
@Service()
export class SubmissionController {
  @Get('/')
  public async postSubmissions() {
    return videoReview
  }
}
