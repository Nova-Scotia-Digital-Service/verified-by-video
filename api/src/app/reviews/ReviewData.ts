import * as TD from '../../types'

import { faker } from '@faker-js/faker'

import { pool } from '../../db'
import { buildNestedPgParams, buildPgParams } from '../../utils/buildPgParams'

const generateReviewQuestions = () => {
  const fakePerson = {
    firstName: faker.person.firstName().toUpperCase(),
    lastName: faker.person.lastName().toUpperCase(),
    driversLicenseNumber: faker.string.numeric(8),
    birthdate: faker.date.birthdate().toLocaleDateString('en-CA', {
      month: 'long',
      day: 'numeric',
      year: 'numeric',
    }),
  }

  const reviewQuestionTemplate: { question: string; options: { text: string }[] }[] = [
    {
      question: 'Did the user correctly follow the prompts in the video?',
      options: [
        { text: 'Yes, all prompts followed by user in the correct order' },
        { text: "Didn't follow the prompts correctly" },
        { text: 'Too low quality video/photo for matching' },
        { text: "Couldn't complete for other reason" },
      ],
    },
    {
      question: 'What name did they provide?',
      options: [
        { text: `${fakePerson.lastName}, ${fakePerson.firstName} or an acceptable variation` },
        { text: "Didn't provide the correct name" },
        { text: "Couldn't complete for other reason" },
      ],
    },
    {
      question: 'Do all three match: photo on the ID document, video, photo taken in the app?',
      options: [
        { text: 'Yes, all match' },
        { text: 'Not all match' },
        { text: 'Too low quality video or photo for matching' },
        { text: "Couldn't complete for other reason" },
      ],
    },
    {
      question: 'What type of ID is provided?',
      options: [{ text: "Nova Scotia Driver's License" }, { text: 'Other type' }, { text: "I can't confirm" }],
    },
    {
      question: "What number is on Nova Scotia Driver's License?",
      options: [{ text: fakePerson.driversLicenseNumber }, { text: 'Other type' }, { text: "I can't confirm" }],
    },
    {
      question: 'What is the birthdate on the ID?',
      options: [{ text: fakePerson.birthdate }, { text: 'Does not match' }, { text: "I can't confirm" }],
    },
    {
      question: 'What name is on the ID?',
      options: [
        { text: `${fakePerson.lastName}, ${fakePerson.firstName} or an acceptable variation` },
        { text: 'Does not match' },
        { text: "I can't confirm" },
      ],
    },
    {
      question: 'Is the ID valid?',
      options: [
        { text: 'Is valid, not expired' },
        { text: 'Expired' },
        { text: "Doesn't appear to be genuine (e.g. not an original document, missing security feature)" },
        { text: 'Not an acceptable document' },
      ],
    },
    {
      question: 'Are you confident this is who they say they are?',
      options: [
        { text: 'Yes I am confident' },
        { text: 'Possible suspicious activity' },
        { text: 'Not confident enough to verify for other reason' },
      ],
    },
  ]
  return reviewQuestionTemplate
}

export const createReview = async (reviewer: TD.DBUser, submission_id: string) => {
  const client = await pool.connect()
  await client.query('BEGIN')
  const reviewResult = await client.query<Pick<TD.DBReview, 'id'>>(
    `
    INSERT INTO reviews (submission_id, reviewer_id)
    VALUES ($1, $2)
    RETURNING id
    `,
    [submission_id, reviewer && reviewer.id],
  )

  const reviewQuestions = generateReviewQuestions()
  await Promise.all(
    reviewQuestions.map(async (question) => {
      const questionResult = await client.query<{ id: string }>(
        `
      INSERT INTO review_questions (review_id, question)
      VALUES ($1, $2)
      RETURNING id

      `,
        [reviewResult.rows[0].id, question.question],
      )

      const reviewQuestionId = questionResult.rows[0].id
      const questionOptionPairs = question.options.map((option) => [reviewQuestionId, option.text])
      await client.query(
        `
      INSERT INTO review_question_options (review_question_id, text)
      VALUES ${buildNestedPgParams(questionOptionPairs)}
      `,
        questionOptionPairs.flat(),
      )
    }),
  )

  await client.query(
    `
    UPDATE submissions
    SET
      status = 'UNDER_REVIEW'
    WHERE id = $1
    `,
    [submission_id],
  )

  await client.query('COMMIT')
  await client.release()

  return reviewResult.rows[0]
}

export const getReview = async (review_id: string) => {
  const client = await pool.connect()
  await client.query('BEGIN')
  const review = await client.query<{
    id: string
    status: TD.ReviewStatus
    video_url: string
    upload_date: Date
    session_id: string
    submission_id: string
    tags: { id: string; text: string }[]
  }>(
    `
    WITH
      aggregated_tags AS (
        SELECT
          submission_tags.submission_id,
          jsonb_build_object(
            'id', tags.id,
            'text', tags.text
          ) AS tags
        FROM submission_tags
        JOIN tags ON tags.id = submission_tags.tag_id
        GROUP BY submission_tags.submission_id, tags.id
      )
    SELECT
      reviews.id,
      reviews.status,
      submissions.id as submission_id,
      submissions.session_id,
      submissions.video_url,
      submissions.upload_date,
      array_remove(
        array_agg(aggregated_tags.tags),
        NULL
      ) AS tags
    FROM reviews
    JOIN submissions ON submissions.id = reviews.submission_id
    LEFT JOIN aggregated_tags ON aggregated_tags.submission_id = submissions.id
    WHERE reviews.id = $1
    GROUP BY submissions.id, reviews.id
    `,
    [review_id],
  )
  const questions = await client.query<{ id: string; question: string; option_id: string; option_text: string }>(
    `
    SELECT
      review_questions.id,
      review_questions.question,
      review_question_options.id as option_id,
      review_question_options.text as option_text
    FROM review_questions
    JOIN review_question_options ON review_questions.id = review_question_id
    WHERE review_id = $1
    `,
    [review_id],
  )
  const selected_options = await client.query<{ selected_option_ids: string[] | null }>(
    `
    SELECT
      array_agg(review_answers.option_id) as selected_option_ids
    FROM review_questions
    JOIN review_question_options ON review_questions.id = review_question_id
    JOIN review_answers ON review_question_options.id = review_answers.option_id
    WHERE review_id = $1
    `,
    [review_id],
  )
  const prompts = await client.query<{ id: string; session_id: string; text: string }>(
    `
    SELECT
      id,
      session_id,
      text
    FROM prompts
    WHERE session_id = $1
    `,
    [review.rows[0].session_id],
  )
  const identification_cards = await client.query<TD.DBIdentificationCard>(
    `
    SELECT
      id,
      session_id,
      description,
      photo_url,
      upload_date
    FROM identification_cards
    WHERE session_id = $1
    `,
    [review.rows[0].session_id],
  )
  await client.query('COMMIT')
  await client.release()

  return {
    review: review.rows[0],
    questions: questions.rows,
    selectedOptions: selected_options.rows[0].selected_option_ids,
    prompts: prompts.rows,
    identificationCards: identification_cards.rows,
  }
}

export const finishReview = async (
  user: TD.DBUser,
  review_id: string,
  status: Exclude<TD.ReviewStatus, 'STARTED'>,
  answers: string[],
  comment?: string,
) => {
  const client = await pool.connect()
  await client.query('BEGIN')
  await client.query(
    `
    INSERT INTO review_answers
      (option_id)
    VALUES
      ${buildPgParams(answers)}
    `,
    answers,
  )
  await client.query(
    `
    UPDATE reviews
    SET
      status = $1,
      reviewer_id = $2
    WHERE id = $3
    `,
    [status, user.id, review_id],
  )
  if (comment) {
    await client.query(
      `
      INSERT INTO review_comments
        (review_id, reviewer_id, text)
      VALUES
        ($1, $2, $3)
      `,
      [review_id, user.id, comment],
    )
  }

  const submissionStatus: TD.SubmissionStatus = status
  await client.query(
    `
    UPDATE submissions
    SET
      status = $2
    FROM reviews
    WHERE
      submissions.id = reviews.submission_id
      AND reviews.id = $1
    `,
    [review_id, submissionStatus],
  )

  await client.query('COMMIT')
  await client.release()
}
