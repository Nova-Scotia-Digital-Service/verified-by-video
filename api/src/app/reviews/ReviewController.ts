import * as TD from '../../types'

import { Get, Controller, Param, UseGuards, Post, Body, Req } from '@nestjs/common'
import { ApiBearerAuth as SwaggerRequireAuth } from '@nestjs/swagger'

import { AuthGuard } from '../auth/AuthGuard'

import { minioClient } from '../../minio'
import config from '../../config'

import { getReview, getReviewList, postReviewAnswers } from './ReviewData'

@Controller('/staff/reviews')
export class ReviewController {
  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Get('/')
  public async getReviewList(): Promise<TD.ReviewSummary[]> {
    return (await getReviewList()).map(
      ({ id, status, tags, submission_id, session_id, video_url, upload_date, reviewer }) => ({
        id,
        status,
        submission: {
          id: submission_id,
          session_id,
          video_url,
          upload_date,
          tags,
        },
        reviewer,
      }),
    )
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Get('/:review_id')
  public async getReview(@Param('review_id') review_id: string): Promise<TD.Review> {
    const [review, questions, prompts, identification_cards] = await getReview(review_id)

    const reviewQuestions: TD.ReviewQuestion[] = []

    for (const question of questions) {
      const existingQuestion = reviewQuestions.find(({ id }) => question.id === id)
      if (!existingQuestion) {
        const newQuestion = {
          id: question.id,
          question: question.question,
          options: [
            {
              id: question.option_id,
              text: question.option_text,
            },
          ],
        }
        reviewQuestions.push(newQuestion)
      } else {
        existingQuestion.options.push({
          id: question.option_id,
          text: question.option_text,
        })
      }
    }

    let signedVideoUrl = await minioClient.presignedGetObject(config.S3_BUCKET_NAME, review.video_url, 60 * 60)
    if (config.NODE_ENV === 'development') {
      signedVideoUrl = signedVideoUrl.replace('http://minio:9000', 'http://localhost:9002')
    }

    const signedIdentificationCards = await Promise.all(
      identification_cards.map(async (card) => {
        if (!card.photo_url) return card

        let signedPhotoUrl = await minioClient.presignedGetObject(config.S3_BUCKET_NAME, card.photo_url, 60 * 60)
        if (config.NODE_ENV === 'development') {
          signedPhotoUrl = signedPhotoUrl.replace('http://minio:9000', 'http://localhost:9002')
        }
        return {
          ...card,
          photo_url: signedPhotoUrl,
        }
      }),
    )

    return {
      id: review.id,
      status: review.status,
      submission: {
        id: review.submission_id,
        session_id: review.session_id,
        video_url: signedVideoUrl,
        upload_date: review.upload_date,
        tags: review.tags,
      },
      identification_cards: signedIdentificationCards,
      prompts: prompts.map((prompt) => ({ ...prompt, type: 'text' })),
      questions: reviewQuestions,
    }
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Post('/:review_id')
  public async postReview(
    @Req() request: { user: TD.DBUser },
    @Param('review_id') review_id: string,
    @Body()
    body: { status: Exclude<TD.ReviewStatus, 'STARTED'>; answers: { [question: string]: string }; comment?: string },
  ): Promise<void> {
    await postReviewAnswers(request.user, review_id, body.status, Object.values(body.answers), body.comment)
  }
}
