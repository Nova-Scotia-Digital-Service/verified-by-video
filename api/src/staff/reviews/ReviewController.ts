import { Get, JsonController, Param } from 'routing-controllers'
import { Service } from 'typedi'

import { mockReviewList, mockReview } from './ReviewData'

@JsonController('/staff/reviews')
@Service()
export class ReviewController {
  @Get('/')
  public async getReviewList() {
    return mockReviewList
  }

  @Get('/:reviewId')
  public async getReview(@Param('reviewId') reviewId: string) {
    return mockReview
  }
}
