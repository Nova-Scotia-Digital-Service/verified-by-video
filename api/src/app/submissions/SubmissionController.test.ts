import { Readable } from 'stream'

import { pool } from '../../db'
import { populateDb, unpopulateDb } from '../../../test/testUtils'

import { StaffSubmissionController, SubmissionController } from './SubmissionController'
import { ReviewController } from '../reviews/ReviewController'

const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
const VIDEO_URL_REGEX = /^media\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}.mp4$/
const PHOTO_URL_REGEX = /^media\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}.png$/

beforeAll(async () => {
  await populateDb()
})

afterAll(async () => {
  await unpopulateDb()
  await pool.end()
})

describe('StaffSubmissionController', () => {
  describe('getSubmissionList', () => {
    it('returns expected data', async () => {
      const submissionList = await new StaffSubmissionController().getSubmissionList()
      expect(submissionList).toMatchSnapshot()
    })
  })

  describe('applyTagToSubmission', () => {
    it('applies a tag to a submission', async () => {
      const reviewId = 'cbb5b46a-ad3b-4a5f-954f-bff18024d1d6'
      const tagId = 'fc1bdf3c-6457-4604-aeb6-48ba42a73eb9'

      const reviewBefore = await new ReviewController().getReview(reviewId)
      expect(reviewBefore.submission.tags).toEqual([])

      await new StaffSubmissionController().applyTagToSubmission(reviewBefore.submission.id, { tag_id: tagId })

      const reviewAfter = await new ReviewController().getReview(reviewId)
      expect(reviewAfter.submission.tags).toEqual([{ id: tagId, text: 'Tag 3' }])
    })
  })

  describe('removeTagFromSubmission', () => {
    it('removes a tag from a submission', async () => {
      const reviewId = 'cbb5b46a-ad3b-4a5f-954f-bff18024d1d6'
      const otherReviewId = '532bd3f4-a0c5-4a97-87a9-19d46981747d'
      const tagId = 'fc1bdf3c-6457-4604-aeb6-48ba42a73eb9'

      const reviewBefore = await new ReviewController().getReview(reviewId)
      const otherReviewBefore = await new ReviewController().getReview(otherReviewId)
      expect(reviewBefore.submission.tags).toEqual([{ id: tagId, text: 'Tag 3' }])
      expect(otherReviewBefore.submission.tags.length).toBe(2)

      await new StaffSubmissionController().removeTagFromSubmission(reviewBefore.submission.id, tagId)

      const reviewAfter = await new ReviewController().getReview(reviewId)
      const otherReviewAfter = await new ReviewController().getReview(otherReviewId)
      expect(reviewAfter.submission.tags).toEqual([])
      expect(otherReviewAfter.submission.tags.length).toBe(2)
    })
  })
})

describe('SubmissionController', () => {
  describe('createPhotoID', () => {
    it('posts a photo', async () => {
      const sessionId = '88acfa3d-bf2f-4cae-8746-60a9106f6d56'
      const photo_file: Express.Multer.File = {
        fieldname: 'file',
        originalname: 'file.png',
        encoding: 'utf-8',
        mimetype: 'image/png',
        size: 0,
        stream: new Readable(),
        destination: '',
        filename: '',
        path: '',
        buffer: Buffer.from(''),
      }
      const newPhotoID = await new SubmissionController().createPhotoID(
        { sessionId, description: 'a test file', photo_file: undefined as any },
        photo_file,
      )
      expect(newPhotoID.id).toMatch(UUID_REGEX)
      expect(newPhotoID.session_id).toBe(sessionId)
      expect(newPhotoID.upload_date).toBeInstanceOf(Date)
      expect(newPhotoID.photo_url).toMatch(PHOTO_URL_REGEX)
    })
  })

  describe('createSubmission', () => {
    it('returns data', async () => {
      const sessionId = '88acfa3d-bf2f-4cae-8746-60a9106f6d56'
      const video_file: Express.Multer.File = {
        fieldname: 'file',
        originalname: 'file.mp4',
        encoding: 'utf-8',
        mimetype: 'video/mp4',
        size: 0,
        stream: new Readable(),
        destination: '',
        filename: '',
        path: '',
        buffer: Buffer.from(''),
      }
      const newSubmission = await new SubmissionController().createSubmission(
        { sessionId, video_file: undefined as any },
        video_file,
      )
      expect(newSubmission.id).toMatch(UUID_REGEX)
      expect(newSubmission.session_id).toBe(sessionId)
      expect(newSubmission.upload_date).toBeInstanceOf(Date)
      expect(newSubmission.video_url).toMatch(VIDEO_URL_REGEX)
    })
  })

  describe('getSubmission', () => {
    it('returns data with correct id', async () => {
      const submissionId = '214f747f-778a-4d4b-ab3b-5ed0c9410c69'
      expect(await new SubmissionController().getSubmission(submissionId)).toStrictEqual({
        id: submissionId,
        video_url: '/media/example-video.mp4',
        session_id: '88acfa3d-bf2f-4cae-8746-60a9106f6d56',
        upload_date: new Date('2024-05-08 19:01:48.776913+00'),
      })
    })
  })
})
