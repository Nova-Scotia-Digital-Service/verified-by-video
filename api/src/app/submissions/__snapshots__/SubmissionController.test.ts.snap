// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffSubmissionController getSubmissionList returns expected data 1`] = `
[
  {
    "id": "214f747f-778a-4d4b-ab3b-5ed0c9410c69",
    "reviews": [
      {
        "created_at": 2024-05-09T14:41:23.551Z,
        "id": "532bd3f4-a0c5-4a97-87a9-19d46981747d",
        "reviewer_name": "User One",
        "status": "STARTED",
      },
    ],
    "status": "UNDER_REVIEW",
    "tags": [
      {
        "id": "1529c4ad-c5f7-404c-b880-6ffc9ad4ca1c",
        "text": "Tag 1",
      },
      {
        "id": "67302e01-4197-47e1-b78e-466064ac5691",
        "text": "Tag 2",
      },
    ],
    "upload_date": 2024-05-08T19:01:48.776Z,
  },
  {
    "id": "be5a2859-8c30-4970-978f-ea30b295ad86",
    "reviews": [
      {
        "created_at": 2024-05-10T15:26:08.739Z,
        "id": "cbb5b46a-ad3b-4a5f-954f-bff18024d1d6",
        "reviewer_name": "User Two",
        "status": "REJECTED",
      },
    ],
    "status": "REJECTED",
    "tags": [],
    "upload_date": 2024-05-10T13:05:18.691Z,
  },
]
`;
